astropath/bayesian.py

```diff
@@ -0,0 +1,190 @@
+    Calculate p(w|O_i) for a given galaxy
+
+    Args:
+        r_w (np.ndarray):
+            offset from galaxy center in arcsec
+        theta_half (float):
+            Half-light radius of this galaxy in arcsec
+        theta_prior (dict):
+            Parameters for theta prior
+        scale_half (float):
+
+    Returns:
+        np.ndarray: Probability values; un-normalized
+
+    """
+    p = np.zeros_like(r_w)
+    ok_w = r_w < theta_prior['max']*theta_half
+    if theta_prior['method'] == 'core':
+        # Wolfram
+        norm = theta_half * np.log(theta_prior['max']+1)
+        if np.any(ok_w):
+            p[ok_w] = theta_half / (r_w[ok_w] + theta_half) / norm
+    elif theta_prior['method'] == 'uniform':
+        norm = theta_half * theta_prior['max']
+        if np.any(ok_w):
+            p[ok_w] = 1. / norm
+    elif theta_prior['method'] == 'exp':
+        norm = theta_half*scale_half * (scale_half - (
+                scale_half+theta_prior['max'])*np.exp(-theta_prior['max']/scale_half))
+        if np.any(ok_w):
+            p[ok_w] = (r_w[ok_w] / theta_half) * np.exp(-r_w[ok_w]/(scale_half*theta_half)) / norm
+    else:
+        raise IOError("Bad theta method")
+    #
+    if norm == 0:
+        raise ValueError("You forgot to normalize!")
+    # Return
+    return p
+
+
+def px_Oi(box_radius, frb_coord, eellipse, cand_coords,
+          theta_prior, step_size=0.1, return_grids=False):
+    """
+    Calculate p(x|O_i), the primary piece of the analysis
+
+    Main concept:
+        1. Set an area to analyze
+        2. Discretize it to 0.1"
+        3. Convolve
+
+    Args:
+        box_radius (float):
+            Maximum radius for analysis, in arcsec
+        frb_coord (SkyCoord):
+        eellipse (dict):
+            Error ellipse for the FRB
+            a, b, theta
+        cand_coords (SkyCoord):
+            Coordinates of the candidate hosts
+        theta_prior (dict):
+            Parameters for theta prior
+        step_size (float, optional):
+            Step size for grid, in arcsec
+        return_grids (bool, optional):
+            if True, return the calcualtion grid
+
+    Returns:
+        np.ndarray or tuple: p(x|O_i) values and the grids if return_grids = True
+
+    """
+    # Error ellipse
+    pa_ee = eellipse['theta'] # deg
+    dtheta = 90. - pa_ee  # Place a of ellipse along the x-axis
+    # Set Equinox (for spherical offsets)
+    frb_coord.equinox = cand_coords[0].equinox
+    #
+    ngrid = int(np.round(2*box_radius / step_size))
+    x = np.linspace(-box_radius, box_radius, ngrid)
+    xcoord, ycoord = np.meshgrid(x,x)
+
+    # Build the grid around the FRB (orient semi-major axis "a" on our x axis)
+    l_w = np.exp(-xcoord ** 2 / (2 * eellipse['a'] ** 2)) * np.exp(-ycoord ** 2 / (2 * eellipse['b'] ** 2))
+
+    p_xMis, grids = [], []
+    # TODO -- multiprocess this
+    for icand, cand_coord in enumerate(cand_coords):
+
+        # #####################
+        # l(w) -- 2D Gaussian
```

Comment:
this should actually go a few lines up where `l_w` is created

Reply:
good call